File: src/amata1219/xeflection/Reflect.scala
package amata1219.xeflection

object Reflect {

  def on(target: Any): AnyReflected = target match {
    case clazz: Class[_] => on(clazz, null)
    case clazzName: String => on(Class.forName(clazzName), null)
    case instance => on(instance.getClass, instance)
  }

  def on(clazzName: String, instance: Any): AnyReflected = on(Class.forName(clazzName), instance)

  def on(clazz: Class[_], instance: Any): AnyReflected = new AnyReflected(clazz, instance)

}

File: src/amata1219/xeflection/kotlin/Reflect.kt
package amata1219.xeflection.kotlin

object Reflect {

    fun on(target: Any): AnyReflected = when (target) {
        is String -> on(target, null)
        is Class<*> -> on(target, null)
        else -> on(target.javaClass, target)
    }

    fun on(clazzName: String, instance: Any?): AnyReflected = on(Class.forName(clazzName), instance)

    fun on(clazz: Class<*>, instance: Any?) = AnyReflected(clazz, instance)

}